### ISSUE TITLE
Having trouble with page.type not typing all characters

### ISSUE THREAD

**Issue by Nagisan** [puppeteer/puppeteer#1648]
Having some issues where it seems this.page.type is not working properly and I can't figure out why. It works 100% of the time in one of my tests, but I have a second test doing the same thing on a different page (only difference is the new page has more input fields) and it fails most of the time (but occasionally passes). Here's an example of what I'm doing:

```
await this.page.type('[placeholder="Enter ID"]', myObj.id);
await this.page.waitForSelector(`input[value="${myObj.id}"]`);

await this.page.type('[placeholder="Enter Type"]', myObj.type);
await this.page.waitForSelector(`input[value="${myObj.type}"]`);
```

I have 11 total input fields that follow the above pattern on this page, the page that has never flaked out on me has only 6 input fields.

The problem is page.type doesn't appear to be typing everything in myObj, for example myObj.type contains 10 random words, occasionally puppeteer seems to only type some of the characters contained in myObj.type, it sometimes cuts off in the middle of a word, other times it cuts off at the end of a word. The property that doesn't type out completely is random every run, as is the number of characters it actually types out. It is not a character limit in the database or input field (I can manually type in significantly more without a problem and each time I run it cuts off at a different random point, sometimes as few as 5-6 characters, other times as many as 20 or so characters).

Is there any debugging methods that would help me figure this out? Console logging myObj shows the full text for every property, taking screenshots or running puppeteer outside of headless mode shows that it stops inputting characters at random times.

Additionally I've tried adding a delay option into my page.type calls and it seems to make the issue worse, typing only a single character into my first input field before breaking. It seems like it might be losing focus (when watching it in headless: false mode).

**Comment by ggirodda** [comment]
Hello, I have the same issue on a page wich have two inputs text. The last one is incomplete, I often get 7 characters instead of 8

**Comment by Nagisan** [comment]
A bit more troubleshooting since my initial post, this page is pulling data from a database and Puppeteer is losing focus on the field that doesn't get completed (if I add a chai expect in after each page.type, when it fails to fill out a field completely the screenshot it takes does not have focus on the input field).

I think what may be happening is puppeteer is just running too quickly and it begins filling in the fields, at some point during this the page finishes getting data from the server and refreshes the page (not a browser refresh.....I'm using React & Redux, so the state updates with the fetched data and refreshes the display).

This loss of focus is not ever seen by a user (because they won't select an input as quickly as puppeteer will) so it's only an issue with puppeteer itself, I can force it to work fine if I make puppeteer wait 500ms after landing on the page (which is enough time for DB calls to resolve locally but may not work in my remote environment).

I've also tried waiting for networkidle and what not, but I'm getting to this page with a click action not a network call so I don't think I can easily wait for networkidle.

Any suggestions on how best to work around this issue? I don't like waiting for a static time because it will just introduce test flakiness in different environments.

**Comment by aslushnikov** [comment]
> Any suggestions on how best to work around this issue? I don't like waiting for a static time because it will just introduce test flakiness in different environments.

@Nagisan I'd try a few options here:
- if the page fetches something over network and you know the url it loads, you can subscribe to the [`requestfinished`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#event-requestfinished) event and wait for the request to finish
- if you have access to the website codebase, you might issue a call to `console.timeStamp` with some label, e.g. `console.timeStamp('mynicelabel')`. In puppeteer, you'll get a related [`metrics`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#event-metrics) event with a `'mynicelabel'` title

**Comment by aficiomaquinas** [comment]
Is this the expected behavior? I'm having the same problem. The page I'm trying to type into does not fetch anything over the network, so there is no use in subscribing to anything, afaic. How can I prevent this issue? It seems to me that it does not work as expected.

**Comment by Tzook** [comment]
Here's a patch to overcome this method until it is fixed:
`    await page.evaluate((text) => {
        (document.getElementById('my-input')).value = text;
      }, "text-to-inject");`

**Comment by putuoka** [comment]
i'm facing the same issue. the word isn't complete. i type inside frame(iframe) using frame.type or page.keyboard.type even with delay giving me the same results. 

![image](https://user-images.githubusercontent.com/20880839/37894889-420e31ba-3112-11e8-97d0-6096d6c3398a.png)


the word should be type completely as "sudo su" command

**Comment by UmairTehsin** [comment]
In my case below mentioned was solution
await page.type("Selector", "Value");
await page.keyboard.press('Enter');

**Comment by slava-lu** [comment]
still a problem.

**Comment by UmairTehsin** [comment]
you can also check by
await page.click(Selector);
await page.keyboard.type('text');
await page.kyboard.press('Enter');

**Comment by slava-lu** [comment]
@UmairTehsin I do have workaround as described earlier ` await popup.evaluate((text) => { (document.getElementById('password')).value = text; }, user.password);`
I just wanted to point out that this issue still exist

**Comment by moltar** [comment]
Having a similar problem. Having 2 consecutive `await page.type(...)` (e.g. username and password). I am seeing that Puppeteer does not change focus from username to password in time. Part of the password is typed into the username field.

**Comment by RobinNagpal** [comment]
> Here's a patch to overcome this method until it is fixed:
> `await page.evaluate((text) => { (document.getElementById('my-input')).value = text; }, "text-to-inject");`

This  is not useful in react project as it doesn't trigger onChange handler

**Comment by liorur** [comment]
Ran into the same issue, after playing around for a while i discovered this little hack produces a persistent result
```javascript
const input = await page.$(`input.inputClass`);
await input.press('Backspace');
await input.type(text);
```

**Comment by jsilvia721** [comment]
> > Here's a patch to overcome this method until it is fixed:
> > `await page.evaluate((text) => { (document.getElementById('my-input')).value = text; }, "text-to-inject");`
> 
> This is not useful in react project as it doesn't trigger onChange handler

@RobinNagpal Do you know if there is a way to do this and trigger the onchange handler?

**Comment by oakgary** [comment]
why is this closed
problem still exists

**Comment by vchenin** [comment]
Hi there!

**Comment by vchenin** [comment]
Hi there!

I think this problem can be solved by waiting for element or attribute which will be created by framework (Angular, jComponent, React, Vue, etc.) engine on browser side.

For example, in my case in authentication form exist button for submit which enabled by default. Component disable this button if there are empty required fields when form initialization and pre-validation finished.

```javascript
await page.waitFor('button[name="submit"][disabled]')
```
Therefore, when it exists, it means that form elements are ready for input.

**Comment by vchenin** [comment]
And after typing:
```javascript
await page.type('#email', 'email@example.com')
await page.type('#password', 'secret')
```
waiting for completion of form validation:
```javascript
await page.waitFor('button[name="submit"]:not([disabled])')
```
then process the form:
```javascript
await page.click('button[name="submit"]')
```

**Comment by RobinNagpal** [comment]
This works for me.

```
export async function typeInInputElement(page, inputSelector, text) {
  await page.evaluate((inputSelector, text) => {
    // Refer to https://stackoverflow.com/a/46012210/440432 for the below solution/code
    const inputElement = document.querySelector(inputSelector);
    const nativeInputValueSetter = Object.getOwnPropertyDescriptor(window.HTMLInputElement.prototype, 'value').set;
    nativeInputValueSetter.call(inputElement, text);

    const ev2 = new Event('input', {bubbles: true});
    inputElement.dispatchEvent(ev2);

  }, inputSelector, text);
}
```

**Comment by rienheuver** [comment]
I was having trouble with the page not having fully loaded yet. The fix for me was to add this as second parameter to `page.goto`:
```
{waitUntil: 'networkidle2'}
```
This makes it wait until there have been no more network requests for the last 500ms. So the total `page.goto` looks like this:
```
await page.goto('https://example.com', {waitUntil: 'networkidle2'});
```
Hope this helps someone out :smile:

**Comment by Fetz** [comment]
Another thing that can cause issues with the type, is animations.

If you have any animation being done in the input (directly to the input or a parent element) you should wait until the animation is finished before using the `page.type` or `element.type`.

for example a bootstrap modal:
```javascript
await page.click('[data-target="#createUserDialog"]'); // show the modal
await page.waitForSelector('#createUserDialog', {visible: true})); // wait for the modal to be visible

const animationTime = 800;
await new Promise((resolve) => setTimeout(resolve, animationTime)); // wait for animation to finish
await page.type('#message', 'if the message is longer, the easy is to replicate the issue with the input');
```

**Comment by fringd** [comment]
Seems a lot of people are still seeing this. Can we re-open it?

**Comment by fringd** [comment]
^^ @aslushnikov ^^

**Comment by fringd** [comment]
My own test was not making any network connections at all, and I've disable animations, but seems to still be flaking with random characters missing.

I have multiple tests running in multiple tabs, I think this is part of the problem, since I can't reproduce if I run a single test suite in isolation.

**Comment by Fetz** [comment]
@fringd if the inputs are simple input fields (without any fancy behaviours when typing) you can set directly the input value instead of using `type`

**Comment by Fetz** [comment]
@fringd as well check if you not forgetting to use await in any `type`

**Comment by fringd** [comment]
Okay I'll try to verify this is not a missing await somewhere. Will update

**Comment by fenilkanjani** [comment]
Trying to input same text in two fields with same values (password and confirm password) on a page after `networkidle2`. It does not type all the characters every time making the test flaky. The code looks something like this: 

```
await page.goto('myurl', {waitUntil: 'networkidle2', ...myOtherOptions})
await page.waitFor('.js--new-password')
await page.type('.js--new-password input', 'test@1234')
await page.type('.js--confirm-password input', 'test@1234')
```

@aslushnikov ^ Any suggestions?

**Comment by fringd** [comment]
this looks a lot like the test I have that's failing @fenilkanjani. Mine's a little more complex, so it's harder to isolate.

**Comment by fringd** [comment]
@fenilkanjani are you running multiple tests in parallel like me? Multiple tabs and threads in jest?

**Comment by aslushnikov** [comment]
@fringd @fenilkanjani are you guys running headless or headful? If you have a good repro that I can run and debug locally, feel free to file a separate issue.

**Comment by fringd** [comment]
I'm running headless

**Comment by fringd** [comment]
I'm also filling out multiple text inputs in one tab while controlling multiple pages in multiple threads.

I haven't got a ton of time for a minimal repro atm :(

**Comment by fringd** [comment]
but running jest with "-i" to run the test in serial fixes it. Are we allowed to do other puppeteer stuff in another tab like click on things at the same time as we are awaiting type?

**Comment by fringd** [comment]
man I'm not even sure if -i fixes it. Flaky tests are hard. I know I need to get you guys a minimal reproduction if this is gonna get fixed.

**Comment by fringd** [comment]
Okay if anybody else finds this. the thing that fixed it for me was setting "slowMo: 1"

1ms doesn't significantly affect my test runtime, and seems to kludgily remove my flake.

@aslushnikov maybe this clue gives you an idea, but if not hopefully someone can reproduce minimally soon.

**Comment by fringd** [comment]
Okay so this is actually a bug in our code that was causing characters to drop if typed too fast. Some other tools like cypress throttle keypresses to 1 per 10ms to be more human-like, but this was actually causing perceivable bugs for our users. Just a coincidence i guess that other people saw this same issue

**Comment by aslushnikov** [comment]
> Okay so this is actually a bug in our code that was causing characters to drop if typed too fast. Some other tools like cypress throttle keypresses to 1 per 10ms to be more human-like, but this was actually causing perceivable bugs for our users. Just a coincidence i guess that other people saw this same issue

@fringd nicee, looks like PPTR served you well! :fire:

**Comment by LorenzHenk** [comment]
What worked for me was to use the `slowMo` paramter when launching puppeteer:
```js
const browser = await puppeteer.launch({
  slowMo: 50,
});
```

**Comment by Jesus75** [comment]
8ball44

________________________________

**Comment by kingzurley** [comment]
Ive used these force approaches to sending special characters to an input, hope this helps!

```
await this.page.click(input, { clickCount: 2 })
await this.page.keyboard.sendCharacter('#@%^%#$@#$@#.')
```

or this

```
await this.page.focus(input)
await this.page.keyboard.sendCharacter('#@%^%#$@#$@#.')
```

**Comment by mauricionr** [comment]
`slowMo` is an option in `puppeteer.launch`, it's similar to the option in `page.type` `{delay: 1}`?

https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pagetypeselector-text-options

**Comment by mauricionr** [comment]
i found my issue with one of my flaky tests, in frontend we was changing the focus of some input and this cause an error on jest because pupeeteer continue typing, but in another input. i removed the timeout to change the focused input and the test passed whitout any `sleep` or `slowMo` or `delay`

**Comment by 9oelM** [comment]
Ok I don't get this.

```js
  await page.type("#username", "myusername");
  await page.type("#password", "mypassword");
```

![image](https://user-images.githubusercontent.com/22465806/62017419-8817b300-b1f1-11e9-97ce-cf3ad281bd27.png)

This will correctly fill out input forms as desired. 

But the code below does not work? This will only fill the input field for`#password`. I don't know why. These two code blocks above and below are essentially the same things. What's happening here? The code below will fill the password input form with some string like `mymyuspasswername`, a random mix of `myusername` and `mypassword`.

```js
const loginInfo: LoginInfo = {
        username: {
            selector: "#username",
            value: "myusername",
        },
        password: {
            selector: "#password",
            value: "mypassword",
        },
    };

Object.keys(loginInfo).forEach(async (key: string) => {
        const { selector, value } = loginInfo[key];
        await page.type(selector, value);
    },
  );
```

![image](https://user-images.githubusercontent.com/22465806/62017523-fbb9c000-b1f1-11e9-8468-6e7c961db1db.png)

**Comment by aslushnikov** [comment]
> Ok I don't get this.

@9oelM your code: 

```js
Object.keys(loginInfo).forEach(async (key: string) => {
        const { selector, value } = loginInfo[key];
        await page.type(selector, value);
    },
  );
```
is basically equal to the following (notice the lack of "await" statements):

```js
page.type("#username", "myusername");
page.type("#password", "mypassword");
```

This tries to type into two fields simultaneously; this won't work - page might have focus only in one field.

**Comment by 9oelM** [comment]
@aslushnikov Thank you for answering my question! 

> This tries to type into two fields simultaneously; this won't work - page might have focus only in one field.

Yes! I believe your answer is very probable. 

But what about `await page.type(selector, value)`? Why is this going to be equal to the code without `await`? I'm sorry for my bad understanding... 

Adding onto that, what could I do if I want to make a loop like above with `forEach` that works correctly?

**Comment by 9oelM** [comment]
@aslushnikov 
Oops, found an answer from #1958 and #3152. 
Thanks!!!

**Comment by Remzes** [comment]
You can use `page.keyboard.type`

Here is a simple function which I wrote
```
const typeWord = async (page, word) => {
  const splitWord = word.split('');
  splitWord.forEach(async key => {
    await page.keyboard.press(key)
  });
};
```

You need to pass a word, which you want to be written to `input`, or any element. If it is uppercase letter or special character, it works as well.

Basically, it works like this:
```
await typeWord(page, 'veryveryveryverylongword');
```

**Comment by ericsoco** [comment]
@Tzook's solution got me most of the way there, but in my case I have an `onChange` handler (as a React prop) that I need to fire in my test, and just setting `input.value = text` doesn't cause the `'change'` event to be dispatched.

This is a bit of a hack, but a subsequent `elementHandle.type()` on the last character got me where I needed to be:

```
    async function enterSearchText(searchText: string) {
      const mostOfSearchText = searchText.slice(0, -1);
      const lastChar = searchText.slice(-1);
      const inputHandle = await page.waitForSelector(SELECTORS.searchInput);
      await page.evaluate(
        (inputEl, text) => (inputEl.value = text),
        inputHandle,
        mostOfSearchText
      );
      return await inputHandle.type(lastChar);
    },

```

Note that I have some bits in my closure that you may not, namely `page` and `SELECTORS`; you can use `document.querySelector` instead of passing an `elementHandle` to `page.evaluate` if that works better for your use case.

**Comment by SaurabhAptask** [comment]
> Here's a patch to overcome this method until it is fixed:
> ` await page.evaluate((text) => { (document.getElementById('my-input')).value = text; }, "text-to-inject");`

This is the best solution.. working great for me

**Comment by christopher-francisco** [comment]
Can we re-open this issue? We're getting the same problem. 

```javascript
// `slowMo` is not set on launch options, so I assume it defaults to 0

const element = await page.waitForSelector(selector, { visible: true });
await element.focus();
await element.type(value);
```

Most of the times, the first 1~3 characters are not typed on some inputs. On some rare occasions NOTHING is typed at all on ANY input.

Finally, running it on the Browser context works, but I don't think that's the puppeteer mantra, to do this?
```javascript
await element.evaluate((node, value) => node.value = value, value);
```

**Comment by segevofer** [comment]
> You can use `page.keyboard.type`
> 
> Here is a simple function which I wrote
> 
> ```
> const typeWord = async (page, word) => {
>   const splitWord = word.split('');
>   splitWord.forEach(async key => {
>     await page.keyboard.press(key)
>   });
> };
> ```
> 
> You need to pass a word, which you want to be written to `input`, or any element. If it is uppercase letter or special character, it works as well.
> 
> Basically, it works like this:
> 
> ```
> await typeWord(page, 'veryveryveryverylongword');
> ```

Promise returned from forEach is ignored

```
forEach(async () => {})
```
hence this solution doesn't promise you (literally) the order of chars that will be typed.

**Comment by dougg0k** [comment]
I was having this issue with the first char in every case, solved by doing
```
const input = await page.waitFor("input");
await input.type(str.charAt(0));
await input.type(str);
```

**Comment by Clish254** [comment]
Try to wait for some time before typing
`  await page.focus(selector);
  await page.waitForTimeout(3000);
  await page.type(
    selector,
    value,
    { delay: 100 }
  );
`

**Comment by chunf** [comment]
> Can we re-open this issue? We're getting the same problem.
> 
> ```js
> // `slowMo` is not set on launch options, so I assume it defaults to 0
> 
> const element = await page.waitForSelector(selector, { visible: true });
> await element.focus();
> await element.type(value);
> ```
> 
> Most of the times, the first 1~3 characters are not typed on some inputs. On some rare occasions NOTHING is typed at all on ANY input.
> 
> Finally, running it on the Browser context works, but I don't think that's the puppeteer mantra, to do this?
> 
> ```js
> await element.evaluate((node, value) => node.value = value, value);
> ```

Agree i am still experiencing the same issue here.

**Comment by f0cus10** [comment]
This issue should definitely be reopened. I can confirm that this is happening in my end as well. I had to use the `element.evaluate()` workaround for this to work properly

The workaround also messes with the `slowMo` and the timeout has to be set manually.

**Comment by garcia-marco** [comment]
+1

**Comment by azakero** [comment]
Please reopen the issue. I'm facing the same problem. 

The first couple of letters are not typed most of the time. This is the code I'm working with:

```
await page.waitForSelector('div[role="input"]' );
await page.focus('div[role="input"]');

// finding the index of the blog post which is pending
const pendingPostIndex = await blog.postContent.findIndex( item => item.status === "pending" );

// decoding post content
const textContent = `${Buffer.from(blog.postContent[pendingPostIndex].text, 'base64').toString()}`;

// this gets the full text
console.log('text content of pending post: ', textContent);

// first couple characters are missing most of the time
await page.type('div[role="input"] span', textContent, {
         delay: 100
});
```

Any suggestion on fixing the issue?

**Comment by abd-falyoun** [comment]
Still presented

**Comment by danielcaldas** [comment]
+1 still affected by this one

**Comment by draco90** [comment]
I am also facing this issue

**Comment by azakero** [comment]
> Still presented



> +1 still affected by this one



> I am also facing this issue

Hey guys, try this solution from SO. 

https://stackoverflow.com/questions/67304665/puppeteer-page-type-misses-out-first-couple-of-characters-of-the-string

It worked for me.

**Comment by draco90** [comment]
Hello, I tried all possible ways mentioned in here, my issue doesnt look like it is related to the delay etc, the value in the variable that i am trying to fill the password field with is around 64 characters. And it sometimes types out all 64 but in most cases, it just types out 54 characters, and thats consistent. Is there a limit set? By default?

**Comment by azakero** [comment]
> Hello, I tried all possible ways mentioned in here, my issue doesnt look like it is related to the delay etc, the value in the variable that i am trying to fill the password field with is around 64 characters. And it sometimes types out all 64 but in most cases, it just types out 54 characters, and thats consistent. Is there a limit set? By default?

Hey man, please post a question in StackOverflow with Puppeteer tag. There are some experts in Puppeteer who helped me with some of my issues. You probably will get a faster response there.

**Comment by draco90** [comment]
> Hello, I tried all possible ways mentioned in here, my issue doesnt look like it is related to the delay etc, the value in the variable that i am trying to fill the password field with is around 64 characters. And it sometimes types out all 64 but in most cases, it just types out 54 characters, and thats consistent. Is there a limit set? By default?

Figured out the issue with this, while fetching the arguments via cli and storing it in argv, its for some reason cutting the password argument(which is the last parameter sent via cli) by 10 characters for some reason. I am not sure if the argv array list has some limit set by default? Does anyone know about this?

**Comment by nmonastyrskyi** [comment]
It still reproduces

**Comment by espired** [comment]
+1
Totally still happens

**Comment by brittanyz** [comment]
I am working with a single page react app login (username/password inputs) form. Also still seeing this issue consistently where the first character (or two or three) in the password input gets chopped (or the input isnt visible before puppeteer starts typing). I tried almost everything here: waitForSelector with {visible: true}, adding a delay, waitForNavigation... as one commenter mentioned just setting the value of the input doesnt work because it doesnt trigger an onChange event in react. 

Finally saw a SO post and this worked 
```
    await page.click(`input[name="password"]`, {delay: 100});
    await page.keyboard.press('Backspace');
    await page.type(`input[name="password"]`, password, {delay: 50});
```

But, this is a workaround and I really do this this issue is a bug that should be addressed internally.

**Comment by omeriko** [comment]
+1
Totally still happens

**Comment by remiconnesson** [comment]
It still happens, this is very weird

the code below does works:
```
...
  await page.goto(graph_url);
  await page.waitForNavigation();
  await page.waitForSelector("input[name=email]", {
    visible: true,
    timeout: 15000,
  });
  await page.type('input[name="email"]', email); // this one doesn't type
  await page.type('input[name="password"]', password);
  await page.type('input[name="email"]', email); // workaround
...
```

but the one below doesn't
```
...
  await page.goto(graph_url);
  await page.waitForNavigation();
  await page.waitForSelector("input[name=email]", {
    visible: true,
    timeout: 15000,
  });
  await page.type('input[name="email"]', email); // this one doesn't type
  await page.type('input[name="password"]', password);
...
```

In the code above, we don't get a timeout but the email field doesn't get anything typed in. The password field receives the typed characters.

In the first example, I had to re-add `await page.type('input[name="email"]', email);` to make it work.

But this seems very hacky.

ALSO.

I'm using headless (puppeteer-core)
And this bug doesn't seem to happen when running headful.

**Comment by jonkratz** [comment]
I'm seeing this issue happen at random. No discernible pattern, but `await input.type('email@domain.com')` will work most of the time, but then occasionally only enter `email@d` or `email@domain.` or any random subset of the string. This is on a website with no background data fetching, or input animation (using headless).

Correct me if I'm wrong, but essentially it seems like the Promise that is being `await`-ed doesn't seem to work 100% of the time (or at the least, the Promise is not waiting for the completion of the entered text).

**Comment by SubliemeSiem** [comment]
Why is this closed? This still happens on version 13.1.1. It happens at random, but it still happens often enough to be a problem.

It happens both in headless and headful mode

**Comment by sophiekoonin** [comment]
I'm also finding this. It's making my test completely unreliable as I can't be sure whether it will finish typing the email into the input.

**Comment by Trendy** [comment]
Still happening in the latest version of Puppeteer.

**Comment by Trendy** [comment]
We were struggling with random test failures due to this bug. After some trial and error, this seemed to work more reliably:

```typescript
const typeOnKeyboard = async (page, inputText) => {
  inputText.split('').forEach(async key => {
    await page.keyboard.sendCharacter(key);
  });
};
```

**Comment by uzair004** [comment]
This is weird & inconsistent. Running script for single value was having problem in one field (changing order of characters in the string) by concatenating a space before string it worked.
now running that script n number of times make it even more terrible.

this thing worked for me.
```
await page.$eval(
      "[name='j_idt31:verisysForm:issuedate_input']",
      (InputField, value) => (issueDateField.value = value),
      value
    );
```

**Comment by KMatuszak** [comment]
This still happens in 19.3.0. This workaround works fine for simple pages where onChange doesn't need to be fired:
```ts
await page.evaluate(async (text: string): Promise<void> => {
  (document.querySelector("#inputID") as HTMLInputElement).value = text;
}, text);
```

**Comment by Adarsha58** [comment]
So, the problem is not with the type, but the problem is puppetteer being too fast between the type and any kind of following events. So, we need to ensure the form is properly rerender after type before doing other stuff. 

`const type =  async (page, text, selector) => {
    const prevText = await getInputTextValue(page, selector); // gets the residing text
    await page.type(selector, text); // then type. 

    // waits for the full form re render after type
    const expectedLength = prevText.length + text.length;
    await page.waitForFunction(
        (selector, expectedLength) =>
                document.querySelector(selector).value.length === expectedLength,
        {timeout: 1500},
        selector,
        expectedLength
    );
};`

**Comment by espenrs** [comment]
It's unbelievable that this bug still exists reading all these comments. My fix is sending each character to the input field one by one:
```
var type = value.split('');
type.forEach(function(char) {
    page.type(form.element, char);
});
```

**Comment by farithadnan** [comment]
Still happening with the latest puppeteer

**Comment by jchook** [comment]
+1 also seeing this issue.

As an alternative workaround, you can call `page.focus(selector)` on every keystroke. For example:

```ts
async function pageType(
  page: Page,
  selector: string,
  text: string,
  options?: {
    delay: number;
  }
): Promise<void> {
  for (let ii = 0; ii < text.length; ii++) {
    await page.focus(selector);
    await page.keyboard.type(text[ii], options);
  }
}
```

**Comment by enahs** [comment]
2023 and still facing this bug.

**Comment by jdmiramontes** [comment]
Still seeing this issue, shouldn't be closed at all

**Comment by AbderrahimElhimir** [comment]
The issue is still present

**Comment by lstparrot** [comment]
This is indeed a very strange problem. I solved it by adding a delay when entering characters. Here's an example: 

```javascript
await page.type(selector, typeobject, { delay: 100 })
```

**Comment by uzair004** [comment]
Its probably better not to use this at all, instead get DOM element in page.evaluate & set its value to text you want to enter in text field.

**Comment by Summaw** [comment]
> Its probably better not to use this at all, instead get DOM element in page.evaluate & set its value to text you want to enter in text field.

Unfortunately this doesn't always work especially when typing in a form, the form doesn't recognize that the value is there and will say something like "first name is required" etc.

**Comment by Summaw** [comment]
I am still receiving this same error that everyone above is mentioning, I've tried 3 different ways...
```
 await page.click('#sp_cust_first_name')
 await page.keyboard.type('Aaron')
 await sleep(2000)
```
This does the same thing that @jonkratz metioned above in this comment: [here](https://github.com/puppeteer/puppeteer/issues/1648#issuecomment-953887184)

```
await page.evaluate(async(firstName) => {
  return document.querySelector('#sp_cust_first_name').value = `${firstName}`;
 }, firstName)
```
In this example if gives the problem I mentioned in the comment above, 
"Unfortunately this doesn't always work especially when typing in a form, the form doesn't recognize that the value is there and will say something like "first name is required" etc."

```
await page.type('#sp_cust_first_name', 'Aaron')
```
When I do the 3rd example I gave above, sometimes it will type the entire string out and sometimes it will type 'Aron' or even just 'A' and then move on.
I've also tried just setting a sleep for 2 seconds before typing the in the next selector and unfortunately that also did not work.

I should also mention I am using **puppeteer-extra** instead of **puppeteer-core** but I also see people using puppeteer-core are facing the same issue I still in 2023.

If I find a solution I will post it here.

**Comment by Summaw** [comment]
I've found a temporary solution while using **puppeteer-extra**, I did the following below but I think there should be a permanent solution made to fix this bug.

```javascript
await page.waitForSelector('#sp_cust_first_name', {timeout: 15 * 1000})
await page.focus('#sp_cust_first_name')
await page.keyboard.type(`${firstName}`)
await sleep(2000)
await page.focus('#sp_cust_last_name')
await page.keyboard.type(`${lastName}`)
```

**Comment by Summaw** [comment]
I've also made another way you can do this which is much more accurate for me when typing out each character:
```javascript
await page.waitForSelector('#sp_cust_first_name', {timeout: 10 * 1000})
await sleep(1000)
await page.focus('#sp_cust_first_name')
//await page.keyboard.type('Aaron')
for (let index = 0; index < "Aaron".split('').length; index++) {
  await sleep(300)
  const element = "Aaron".split('')[index];
  await page.keyboard.type(element)  
}
let first_name = await page.evaluate(function(){
  return document.querySelector('#sp_cust_first_name').value;
})
while(first_name !== 'Aaron'){
  await sleep(2000)
  first_name = await page.evaluate(function(){
    return document.querySelector('#sp_cust_first_name').value;
  })
}
```

**Comment by efolio** [comment]
Still facing this issue on 21.5.2… It appears completely random.

**Comment by anshulagarwal92** [comment]
I am also still facing the same issue randomly sometimes it works and sometimes fails, can we please reopen this issue

**Comment by efolio** [comment]
I found a way out of it !
Turns out I wasn’t writing in a real input, but inside a span that was styled as an input ! It wasn’t easy to see…

The issue was that when Puppeteer wrote a character inside the span, it was processed by some JS on the front-end that, by nature, was asynchronous, and, depending on how fast it executed, the `await page.keyboard.type(message)` already returned, and my function already clicked on the submit button, and thus sent only part of the input content !

I solved all my typing issues making a wrapper which does this :

```js
await this.page.click(selector)
await this.page.keyboard.type(message)
await this.page.waitForFunction(message => document.querySelector(selector).innerText === message, {}, message)
```

This last line forces the execution to wait until the JS behind is executed and the content is properly rendered :)

**Comment by navigatrum** [comment]
Basing on https://github.com/puppeteer/puppeteer/issues/1648#issuecomment-1864851917 I'm using:
```ts
import { Page, KeyboardTypeOptions } from 'puppeteer'

async function safeType(
    page: Page,
    selector: string,
    text: string,
    options?: Readonly<KeyboardTypeOptions>): Promise<void> {
    await page.type(selector, text, options)
   
    /** wait for the `input` or `textarea` element `value` to be equal to the fully typed text */
    await page.waitForFunction((text, selector) =>
        (document.querySelector(selector) as HTMLInputElement || HTMLTextAreaElement).value === text,
        {}, text, selector)
}
```
which would seem to work (`input.innetText` doesn't seem to change even after value has changed)

**Comment by efolio** [comment]
If you write inside an input element or a textarea, you need the value. Otherwise, the innerText ;)

**Comment by ivanalemunioz** [comment]
If any of you are having a hard time debugging Puppeteer, I'm creating a new service to make it easier: [Buglesstack](https://buglesstack.com/). It's currently in beta, so I'd be glad if any of you try it out!

**Comment by whale4113** [comment]
Try this

```typescript
const keyboardInput = async (
  page: puppeteer.Page,
  selector: string,
  value: string,
  options: {
    /**
     * @default true
     */
    waitForReact?: boolean;
  } = {}
) => {
  const { waitForReact = true } = options;

  let current = "";
  for (const char of value) {
    await page.keyboard.sendCharacter(char);

    current += char;
    if (waitForReact) {
      await page.waitForFunction(
        ({ selector, current }) => {
          return document.querySelector<HTMLInputElement>(selector)?.value === current;
        },
        { timeout: 3 * Second },
        { selector, current }
      );
    }
  }
};
```

**Comment by igorshandrov** [comment]
This issue been around since 2017. Just throw away that framework and use a good alternative like playwright. Problem solved.

**Comment by sputnikW** [comment]
use `delay`option solve my problems

``` typescript
await page.type("xxx", text, {
    delay: 100 // delay after type each charater
});
```

**Comment by Summaw** [comment]
> use `delay`option solve my problems
> 
> await page.type("xxx", text, {
>     delay: 100 // delay after type each charater
> });

When I had this issue awhile back, this did not fix the issue.

**Comment by ivanalemunioz** [comment]
Hi, everyone. I’ve been working on a solution to visually debug my browser automation crashes, and I’ve now made it free for the public to use: https://buglesstack.com

I don’t mean to spam or bother you; I just want to share this solution. If you have any questions or suggestions for improvement, feel free to contact me at ivan@buglesstack.com

Thanks for reading! ❤️

**Comment by youngjunning** [comment]
> Here's a patch to overcome this method until it is fixed: ` await page.evaluate((text) => { (document.getElementById('my-input')).value = text; }, "text-to-inject");`

```js
await page.$eval(args.selector, (el, value) => (el as HTMLInputElement).value = value, "1223")
```
